components/datetime/src/metazone.rs

```diff
@@ -0,0 +1,130 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::provider::time_zones::{MetaZoneId, TimeZoneBcp47Id};
+
+use crate::error::DateTimeFormatError;
+use crate::provider::time_zones::MetaZonePeriodV1Marker;
+use icu_calendar::DateTime;
+use icu_calendar::Iso;
+use icu_locid::Locale;
+use icu_provider::prelude::*;
+use zerovec::ule::AsULE;
+
+pub struct MetaZoneCalculator {
+    pub(super) metazone_period: DataPayload<MetaZonePeriodV1Marker>,
+}
+
+/// [`MetaZoneCalculator`] uses data from the [data provider], the selected [`Locale`] to calculate
+/// metazone id.
+impl MetaZoneCalculator {
+    /// Constructor that loads data before calculating metazone id.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_locid::locale;
+    /// use icu::datetime::metazone::MetaZoneCalculator;
+    ///
+    /// let provider = icu_testdata::get_provider();
+    /// let mzc = MetaZoneCalculator::new(
+    ///     locale!("en"),
+    ///     &provider,
+    /// );
+    ///
+    /// assert!(mzc.is_ok());
+    /// ```
+    pub fn new<L, ZP>(locale: L, zone_provider: &ZP) -> Result<Self, DateTimeFormatError>
+    where
+        L: Into<Locale>,
+        ZP: ResourceProvider<MetaZonePeriodV1Marker> + ?Sized,
```

Comment:
Observation: is the data for `MetaZonePeriodV1Marker` the same for all locales? If so, we should probably put it under `und` and include it only one time in the data file. If this is true, please fix it in a subsequent PR.

Reply:
good catch! It looks like they are the same for different locales. Will do it in a separate PR.